apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts

```diff
@@ -92,10 +105,47 @@ export class WorkflowQueueService {
     };
   }
 
+  private async jobHasCompleted(job): Promise<void> {
+    await this.setJobAsCompleted.execute(
+      SetJobAsCommand.create({
+        environmentId: job.data._environmentId,
+        _jobId: job.data._id,
+        organizationId: job.data._organizationId,
+      })
+    );
+  }
+
+  private async jobHasFailed(job, error): Promise<void> {
+    const hasToBackoff = this.runJob.shouldBackoff(error);
+
+    if (!hasToBackoff) {
+      await this.setJobAsFailed.execute(
+        SetJobAsFailedCommand.create({
+          environmentId: job.data._environmentId,
+          error,
+          _jobId: job.data._id,
+          organizationId: job.data._organizationId,
+        })
+      );
+    }
+
+    const lastWebhookFilterRetry = job.attemptsMade === this.DEFAULT_ATTEMPTS && hasToBackoff;
+
+    if (lastWebhookFilterRetry) {
+      await this.handleLastFailedWebhookFilter(job, error);
+    }
```

Comment:
Once understanding what's happening here and found the proper naming, this should be moved to an use case to finalise the domain logic extraction.

Reply:
All he was trying to do is, not to fail the job when number of attempts  not reached(number of attempts 3 for jobs with webhook filters), so those jobs should skip the failure path for first and 2nd failure and on 3rd failure(last failure) it should go to regular failure path and decide to schedule next job or no, based on the flag shouldStopOnFailt.. this is also taken care very cleanly in #2727 and much simpler to understand in new code